### PR TITLE
Scripting solution

### PR DESCRIPTION
**Pull request by anatoliykmetyuk**
This PR adds scripting support to Scala 3. Given a source named `Test.scala`:

```scala
@main def Test(name: String): Unit =
  println(s"Hello ${name}!")
```

You can run: `./bin/scala ../playground/foo/Test.scala World` to get an output `Hello World`.

Internally, `scala` detects the first residual argument which ends with `*.scala`. It then assumes it to be a file that the user wants to run as a script, and all the arguments after that file are treated as the arguments to that script.

A "script" is an ordinary Scala file which contains a `@main` method. The semantics of the `scala Script.scala` command are as follows:

- Compile `Script.scala` with `scalac` into a temporary directory
- Detect the main method in the `*.class` files produced by the compilation
- Execute the main method reflectively

### CODE REVIEW

File: compiler/src/dotty/tools/scripting/ScriptingDriver.scala [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    val candidates = for
+      file <- outDir.toFile.listFiles.toList
+      method <- collectMainMethods(file, "")
+    yield method
+
+    candidates match
+      case Nil =>
+        throw ScriptingException("No main methods detected in your script")
+      case _ :: _ :: _ =>
+        throw ScriptingException("More than one main method detected in your script")
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/scripting/ScriptingDriver.scala:73]: We should tell which ones were found.

File: compiler/test/dotty/tools/scripting/ScriptingTests.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      scriptFile <- testFiles
+      if scriptFile.getName.endsWith(".scala")
+      name = scriptFile.getName.dropExtension
+      scriptArgs = argss.getOrElse(name, Array.empty[String])
+    do
+      ScriptingDriver(
+        compilerArgs = Array(
+            "-classpath", TestConfiguration.basicClasspath,
+            "-color:never",
+            "-Yerased-terms",
```
Comment (nicolasstucki) [inline compiler/test/dotty/tools/scripting/ScriptingTests.scala:38]: Why did you need this one? I feel wrong to add it by default, users should opt-in to use this feature.

File: compiler/src/dotty/tools/scripting/ScriptingDriver.scala [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import dotty.tools.dotc.config.CompilerCommand
+import dotty.tools.io.{ PlainDirectory, Directory }
+import dotty.tools.dotc.reporting.Reporter
+import dotty.tools.dotc.config.Settings.Setting._
+
+import sys.process._
+
+class ScriptingDriver(compilerArgs: Array[String], scriptFile: File, scriptArgs: Array[String]) extends Driver:
+  def compileAndRun(): Unit =
+    val outDir = Files.createTempDirectory("scala3-scripting")
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/scripting/ScriptingDriver.scala:21]: We should delete this directory after the script finishes

File: compiler/src/dotty/tools/scripting/ScriptingDriver.scala [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        yield membersMainMethod
+      else if target.getName.endsWith(".class") then
+        val cls = cl.loadClass(targetPath)
+        val method =
+          try cls.getMethod("main", classOf[Array[String]])
+          catch
+            case _: java.lang.NoSuchMethodException => null
+
+        if method != null && Modifier.isStatic(method.getModifiers) then List(method)
+        else Nil
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/scripting/ScriptingDriver.scala:60]: We can avoid the `null` with this refacroring
```scala
try 
  val method = cls.getMethod("main", classOf[Array[String]])
  if Modifier.isStatic(method.getModifiers) then List(method) else Nil
catch
  case _: java.lang.NoSuchMethodException => Nil
```

File: compiler/src/dotty/tools/scripting/ScriptingDriver.scala [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      else Nil
+    end collectMainMethods
+
+    val candidates = for
+      file <- outDir.toFile.listFiles.toList
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/scripting/ScriptingDriver.scala:65]:
```suggestion
      file <- outDir.toFile.listFiles.toIterator
```
or some other way to get the iterator from java to scala

**Review by nicolasstucki — CHANGES_REQUESTED** [review]

**Comment by nicolasstucki** [comment]
We are also missing some docs.

**Review by nicolasstucki — APPROVED** [review]

**Comment by philwalk** [comment]
I have used scala for several years, and make heavy use of scala scripting.   I am eager to contribute to this and have some ideas to greatly improve the user experience for new users.

**Comment by som-snytt** [comment]
This will make it much easier to run quick test comparisons between Scala 2 and 3!

**Comment by rjolly** [comment]
I have a dream that `scala` is just a shorthand for
```scala
jrunscript -cp lib/scala3-compiler_3.0.0-M3.jar:lib/scala3-library_3.0.0-M3.jar:lib/scala3-interfaces-3.0.0-M3.jar:lib/tasty-core_3.0.0-M3.jar:lib/scala-library-2.13.4.jar:lib/scala-asm-7.3.1-scala-1.jar -Djava.class.path=lib/scala3-library_3.0.0-M3.jar:lib/scala-library-2.13.4.jar -l scala
```

**Pull request merged**